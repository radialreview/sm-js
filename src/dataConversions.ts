export const JSON_TAG = '__JSON__';
export const NULL_TAG = '__NULL__';

export function parseJSONFromBE(jsonString: string) {
  if (!jsonString.startsWith(JSON_TAG)) {
    throw Error(`parseJSONFromBE - invalid json received:\n${jsonString}`);
  }

  // convert string array into js array
  if (jsonString.startsWith(`${JSON_TAG}[`)) {
    console.log('NOLEY JSON STRING ERROR', jsonString);
    return JSON.parse(jsonString.replace('__JSON__', ''));
  }

  // Allow new line text (\n to \\n)
  // replacing prevents JSON.parse to complaining
  return JSON.parse(jsonString.replace(JSON_TAG, '').replace(/\n/g, '\\n'));
}

export function prepareValueForFE(value: any): any {
  if (value === NULL_TAG) {
    return null;
  } else if (value === 'true' || value === 'false') {
    return value === 'true';
  } else if (typeof value === 'string' && value.startsWith(JSON_TAG)) {
    return parseJSONFromBE(value);
  } else if (Array.isArray(value)) {
    return value.map(entry => {
      if (typeof entry === 'object') {
        return prepareValueForFE(entry);
      } else {
        return entry;
      }
    });
  } else if (value != null && typeof value === 'object') {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return prepareForFE(value);
  } else {
    return value;
  }
}

export function prepareForFE(beData: Record<string, any>) {
  return Object.keys(beData).reduce((prepared, key) => {
    const value = beData[key];
    return {
      ...prepared,
      [key]: prepareValueForFE(value),
    };
  }, {} as Record<string, any>);
}
