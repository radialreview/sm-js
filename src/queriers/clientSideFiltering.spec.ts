import {
  DEFAULT_TOKEN_NAME,
  EPaginationFilteringSortingInstance,
  MMGQL,
  queryDefinition,
} from '..';
import { NULL_TAG } from '../dataConversions';
import {
  createMockQueryDefinitions,
  generateUserNode,
  getMockConfig,
  mockTodoData,
  mockUserData,
  createMockDataItems,
} from '../specUtilities';

test(`query.filter can filter 'number' prop using 'gte' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          score: 10,
        },
        {
          score: 20,
        },
        {
          score: 30,
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ score }) => ({
        score,
      }),
      filter: {
        score: { gte: 20 },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(2);
});

test(`query.filter can filter 'number' prop using 'lte' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          score: 10,
        },
        {
          score: 20,
        },
        {
          score: 30,
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ score }) => ({
        score,
      }),
      filter: {
        score: { lte: 20 },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(2);
});

test(`query.filter can filter 'number' prop using 'eq' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          score: 10,
        },
        {
          score: 20,
        },
        {
          score: 10,
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ score }) => ({
        score,
      }),
      filter: {
        score: { eq: 10 },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(2);
});

test(`query.filter can filter 'number' prop using 'neq' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          score: 10,
        },
        {
          score: 20,
        },
        {
          score: 30,
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ score }) => ({
        score,
      }),
      filter: {
        score: { neq: 10 },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(2);
});

test(`query.filter can filter 'number' prop using 'gt' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          score: 10,
        },
        {
          score: 20,
        },
        {
          score: 30,
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ score }) => ({
        score,
      }),
      filter: {
        score: { gt: 10 },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(2);
});

test(`query.filter can filter 'number' prop using 'lt' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          score: 10,
        },
        {
          score: 20,
        },
        {
          score: 30,
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ score }) => ({
        score,
      }),
      filter: {
        score: { lt: 20 },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(1);
});

test(`query.filter can filter 'boolean' prop using 'eq' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          archived: true,
        },
        {
          archived: true,
        },
        {
          archived: false,
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ score, archived }) => ({
        score,
        archived,
      }),
      filter: {
        archived: { eq: true },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(2);
});

test(`query.filter can filter 'null' values with 'eq' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          optionalProp: null,
        },
        {
          optionalProp: NULL_TAG,
        },
        {
          optionalProp: NULL_TAG,
        },
        {
          optionalProp: 'withvalue',
        },
      ],
    }),
  });

  expect(
    (
      await mmGQLInstance.query({
        users: queryDefinition({
          def: generateUserNode(mmGQLInstance),
          map: ({ score, optionalProp }) => ({
            score,
            optionalProp,
          }),
          filter: {
            optionalProp: { eq: null },
          },
        }),
      })
    ).data.users.nodes.length
  ).toBe(3);
});

test(`query.filter can filter 'null' values with 'neq' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          optionalProp: NULL_TAG,
        },
        {
          optionalProp: NULL_TAG,
        },
        {
          optionalProp: 'withvalue',
        },
      ],
    }),
  });

  expect(
    (
      await mmGQLInstance.query({
        users: queryDefinition({
          def: generateUserNode(mmGQLInstance),
          map: ({ score, optionalProp }) => ({
            score,
            optionalProp,
          }),
          filter: {
            optionalProp: { neq: null },
          },
        }),
      })
    ).data.users.nodes.length
  ).toBe(1);
});

test(`query.filter can filter 'boolean' prop using 'neq' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          archived: true,
        },
        {
          archived: true,
        },
        {
          archived: false,
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ archived }) => ({
        archived,
      }),
      filter: {
        archived: { neq: false },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(2);
});

test(`query.filter can filter 'string' prop using 'eq' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          firstName: 'John',
        },
        {
          firstName: 'Doe',
        },
        {
          firstName: 'Mary',
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ firstName }) => ({
        firstName,
      }),
      filter: {
        firstName: { eq: 'John' },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(1);
});

test(`query.filter can filter 'string' prop using 'contains' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          firstName: 'John Patrick',
        },
        {
          firstName: 'Patrick John',
        },
        {
          firstName: 'Mary',
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ firstName }) => ({
        firstName,
      }),
      filter: {
        firstName: { contains: 'John' },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(2);
});

test(`query.filter can filter 'string' prop using 'ncontains' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          firstName: 'John Patrick',
        },
        {
          firstName: 'Patrick John',
        },
        {
          firstName: 'Mary',
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ firstName }) => ({
        firstName,
      }),
      filter: {
        firstName: { ncontains: 'John' },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(1);
});

test(`query.filter can filter 'string' prop using 'neq' operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          firstName: 'John',
        },
        {
          firstName: 'John',
        },
        {
          firstName: 'Mary',
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ firstName }) => ({
        firstName,
      }),
      filter: {
        firstName: { neq: 'John' },
      },
    }),
  });

  expect(data.users.nodes.length).toBe(1);
});

test(`query.filter supports old filter object format with 'eq' as default operator`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          firstName: 'John',
        },
        {
          firstName: 'Test',
        },
        {
          firstName: 'Test',
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ firstName }) => ({
        firstName,
      }),
      filter: {
        firstName: 'Test',
      },
    }),
  });

  expect(data.users.nodes.length).toBe(2);
});

test(`query.filter can filter relational data`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          todos: createMockDataItems({
            sampleMockData: mockTodoData,
            items: [
              {
                task: 'My Todo 1',
              },
              {
                task: 'My Todo 2',
              },
              {
                task: 'Todo 8',
              },
            ],
          }),
        },
        {
          todos: createMockDataItems({
            sampleMockData: mockTodoData,
            items: [
              {
                task: 'My Todo 3',
              },
              {
                task: 'Todo 4',
              },
            ],
          }),
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ score, todos }) => ({
        score,
        todos: todos({
          map: ({ task }) => ({ task }),
          filter: {
            task: { contains: 'my todo' },
          },
        }),
      }),
    }),
  });

  expect(data.users.nodes[0].todos.nodes.length).toBe(2);
  expect(data.users.nodes[1].todos.nodes.length).toBe(1);
});

test(`query.filter can filter multilevel relational data`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          todos: createMockDataItems({
            sampleMockData: mockTodoData,
            items: [
              {
                task: 'My Todo 1',
                users: createMockDataItems({
                  sampleMockData: mockUserData,
                  items: [
                    {
                      firstName: 'John',
                    },
                    {
                      firstName: 'Mark',
                    },
                  ],
                }),
              },
            ],
          }),
        },
      ],
    }),
  });

  const { data } = await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ score, todos }) => ({
        score,
        todos: todos({
          map: ({ users }) => ({
            users: users({
              map: ({ firstName }) => ({ firstName }),
              filter: {
                firstName: { contains: 'john' },
              },
            }),
          }),
        }),
      }),
    }),
  });

  expect(data.users.nodes[0].todos.nodes[0].users.nodes.length).toBe(1);
});

test(`query.filter can filter nested object property`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          address__dot__state: 'FL',
        },
        {
          address__dot__state: 'NY',
        },
        {
          address__dot__state: 'FL',
        },
      ],
    }),
  });

  expect(
    (
      await mmGQLInstance.query({
        users: queryDefinition({
          def: generateUserNode(mmGQLInstance),
          map: ({ address }) => ({
            address: address({
              map: ({ state }) => ({ state }),
            }),
          }),
          filter: {
            address: {
              state: { eq: 'FL' },
            },
          },
        }),
      })
    ).data.users.nodes.length
  ).toBe(2);
});

test(`query.filter should throw an error if property being filtered is not defined in the queryDefinition map function`, async () => {
  const { mmGQLInstance } = setupTest({
    users: createMockDataItems({
      sampleMockData: mockUserData,
      items: [
        {
          score: 10,
        },
        {
          score: 20,
        },
        {
          score: 30,
        },
      ],
    }),
  });

  const result = await mmGQLInstance
    .query({
      users: queryDefinition({
        def: generateUserNode(mmGQLInstance),
        map: ({ address }) => ({
          address: address({
            map: ({ state }) => ({ state }),
          }),
        }),
        filter: {
          score: { gte: 20 },
          address: { state: { eq: 'test' } },
        },
      }),
    })
    .catch(e => (e as Error).stack || '');

  expect(result).toContain(
    `FilterPropertyNotDefinedInQueryException exception - The filter property 'score' is not defined in the 'map' function of the queryDefinition. Add that property to the queryDefinition 'map' function`
  );
});

function setupTest(mockData?: any) {
  const mmGQLInstance = new MMGQL(
    getMockConfig({
      mockData: mockData,
      generateMockData: false,
      paginationFilteringSortingInstance:
        EPaginationFilteringSortingInstance.CLIENT,
    })
  );
  mmGQLInstance.setToken({
    tokenName: DEFAULT_TOKEN_NAME,
    token: 'mock token',
  });
  const queryDefinitions = createMockQueryDefinitions(mmGQLInstance);

  return { mmGQLInstance, queryDefinitions, createMockQueryDefinitions };
}
