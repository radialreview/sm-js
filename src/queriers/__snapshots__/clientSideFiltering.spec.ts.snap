// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query.filter accepts undefined as a value 1`] = `
"query query0 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter accepts undefined as a value for a relational query's filter 1`] = `
"query query1 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     todos: todos {
       nodes {
         id
         version
         lastUpdatedBy
         type
         task
       }
       pageInfo {
         endCursor
         startCursor
         hasNextPage
         hasPreviousPage
       }
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'boolean' prop using 'eq' operator 1`] = `
"query query8 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
     archived
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'boolean' prop using 'neq' operator 1`] = `
"query query11 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     archived
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'null' values with 'eq' operator 1`] = `
"query query9 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
     optionalProp
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'null' values with 'neq' operator 1`] = `
"query query10 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
     optionalProp
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'number' prop using 'eq' operator 1`] = `
"query query4 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'number' prop using 'gt' operator 1`] = `
"query query6 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'number' prop using 'gte' operator 1`] = `
"query query2 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'number' prop using 'lt' operator 1`] = `
"query query7 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'number' prop using 'lte' operator 1`] = `
"query query3 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'number' prop using 'neq' operator 1`] = `
"query query5 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'string' prop using 'contains' operator 1`] = `
"query query13 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'string' prop using 'eq' operator 1`] = `
"query query12 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'string' prop using 'ncontains' operator 1`] = `
"query query14 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter 'string' prop using 'neq' operator 1`] = `
"query query15 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter multilevel relational data 1`] = `
"query query18 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
     todos: todos {
       nodes {
         id
         version
         lastUpdatedBy
         type
         users: users {
           nodes {
             id
             version
             lastUpdatedBy
             type
             firstName
           }
           pageInfo {
             endCursor
             startCursor
             hasNextPage
             hasPreviousPage
           }
         }
       }
       pageInfo {
         endCursor
         startCursor
         hasNextPage
         hasPreviousPage
       }
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter multiple relational data 1`] = `
"query query32 {
 user: user(id: \\"mock-user-id\\") {
   id
   version
   lastUpdatedBy
   type
   archived
   todos: todos {
     nodes {
       id
       version
       lastUpdatedBy
       type
       task
       numberProp
     }
     pageInfo {
       endCursor
       startCursor
       hasNextPage
       hasPreviousPage
     }
   }
   otherTodos: todos {
     nodes {
       id
       version
       lastUpdatedBy
       type
       task
       numberProp
     }
     pageInfo {
       endCursor
       startCursor
       hasNextPage
       hasPreviousPage
     }
   }
 }
}"
`;

exports[`query.filter can filter nested object property 1`] = `
"query query19 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     address {
       state
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter query with "AND" condition using the node's nonPaginatedOneToMany relational properties 1`] = `
"query query24 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
     todos: nonPaginatedTodos {
       id
       version
       lastUpdatedBy
       type
       task
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter query with "AND" condition using the node's oneToMany relational properties 1`] = `
"query query22 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
     todos: todos {
       nodes {
         id
         version
         lastUpdatedBy
         type
         task
       }
       pageInfo {
         endCursor
         startCursor
         hasNextPage
         hasPreviousPage
       }
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter query with "AND" condition using the node's oneToOne relational properties 1`] = `
"query query28 {
 todos: todos {
   nodes {
     id
     version
     lastUpdatedBy
     type
     task
     assignee: assignee {
       id
       version
       lastUpdatedBy
       type
       firstName
       archived
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter query with "OR" condition using the node's nonPaginatedOneToMany relational properties 1`] = `
"query query25 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
     todos: nonPaginatedTodos {
       id
       version
       lastUpdatedBy
       type
       task
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter query with "OR" condition using the node's oneToMany relational properties 1`] = `
"query query23 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
     todos: todos {
       nodes {
         id
         version
         lastUpdatedBy
         type
         task
       }
       pageInfo {
         endCursor
         startCursor
         hasNextPage
         hasPreviousPage
       }
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter query with "OR" condition using the node's oneToOne relational properties 1`] = `
"query query27 {
 todos: todos {
   nodes {
     id
     version
     lastUpdatedBy
     type
     task
     assignee: assignee {
       id
       version
       lastUpdatedBy
       type
       firstName
       archived
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter relational data 1`] = `
"query query17 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
     todos: todos {
       nodes {
         id
         version
         lastUpdatedBy
         type
         task
       }
       pageInfo {
         endCursor
         startCursor
         hasNextPage
         hasPreviousPage
       }
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter relational data of a single node query 1`] = `
"query query31 {
 user: user(id: \\"mock-user-id\\") {
   id
   version
   lastUpdatedBy
   type
   archived
   todos: todos {
     nodes {
       id
       version
       lastUpdatedBy
       type
       task
       numberProp
     }
     pageInfo {
       endCursor
       startCursor
       hasNextPage
       hasPreviousPage
     }
   }
 }
}"
`;

exports[`query.filter can filter relational data using "OR" condition 1`] = `
"query query30 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     todos: todos {
       nodes {
         id
         version
         lastUpdatedBy
         type
         task
         numberProp
       }
       pageInfo {
         endCursor
         startCursor
         hasNextPage
         hasPreviousPage
       }
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter can filter using "OR" condition 1`] = `
"query query29 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
     score
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter should throw an error if property being filtered is not defined in the queryDefinition map function 1`] = `
"query query21 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     address {
       state
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter should throw an error if relational prop is not defined in the map function when filtering by relational data 1`] = `
"query query26 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
     todos: todos {
       nodes {
         id
         version
         lastUpdatedBy
         type
       }
       pageInfo {
         endCursor
         startCursor
         hasNextPage
         hasPreviousPage
       }
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter supports old filter object format with 'eq' as default operator 1`] = `
"query query16 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query.filter undefined filters should not be included 1`] = `
"query query20 {
 users: users {
   nodes {
     id
     version
     lastUpdatedBy
     type
     score
     firstName
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;
